File: backend/src/Application/Preferences/ClientPreferences.cs
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

namespace MailFathom.Application.Preferences;

/// <summary>What one person set about their own client, which the deployment holds so it follows them between machines.</summary>
/// <param name="TelemetryEnabled">Whether this deployment may be told what the person's client is doing.</param>
/// <param name="Theme">What the client is painted in once a session exists.</param>
/// <param name="OpenMailInTabs">Whether opening a message opens a tab rather than replacing what is on the screen.</param>
/// <param name="MarkReadOnOpen">Whether opening a message in the client marks it read on the user's own mail server.</param>
/// <param name="ExpandWholeThread">Whether opening a conversation draws every message in it rather than the one it was opened at.</param>
/// <param name="MessageView">Which of the three renderings an open message is drawn on.</param>
/// <param name="AiFiltersShown">Whether the folder tree carries the standing views of what a derivation read in the mail.</param>
/// <param name="NotificationSeconds">How long one of the client's own notifications stands before it takes itself away.</param>
/// <remarks>
/// <para>
/// A closed set of eight rather than a settings service. Each of them says how somebody wants to work rather than what
/// the screen in front of them is like, which is why they belong to the person and not to the browser profile or the
/// desktop install they happened to set them in — and why a ninth is added when there is a ninth to add.
/// </para>
/// <para>
/// Marking read is here rather than on the mail account for the reason
/// <see href="https://github.com/Krzysztof318/MailFathom/blob/main/docs/decisions/0026-marking-a-message-read-when-a-person-opens-it-in-the-client.md">ADR 0026</see>
/// gives: read state is what must not fragment between the machines one person reads on, so it covers every account
/// that user reads and is not an operator's key.
/// </para>
/// <para>
/// It is deliberately not part of the user record. That document is configuration, binds strictly against the rules a
/// configuration file does, and is written under a grant that decides which mailboxes this deployment reads; none of
/// those has anything to do with whether a person may turn telemetry off or choose a theme.
/// </para>
/// <para>
/// Nothing here is personal data about a third party and nothing here is mail. What it does carry is a decision about
/// what may be said about this person, which is why the switch is theirs to set under the grant they already hold
/// rather than under one an administrator maintains for them.
/// </para>
/// </remarks>
public sealed record ClientPreferences(
    bool TelemetryEnabled,
    ClientThemeChoice Theme,
    bool OpenMailInTabs,
    bool MarkReadOnOpen,
    bool ExpandWholeThread,
    ClientMessageView MessageView,
    bool AiFiltersShown,
    int NotificationSeconds)
{
    /// <summary>The shortest a notification may be asked to stand for.</summary>
    /// <remarks>A second is long enough to read a title and reach the control on it, and shorter than that is a notification nobody can act on rather than a preference.</remarks>
    public const int ShortestNotificationSeconds = 1;

    /// <summary>The longest a notification may be asked to stand for.</summary>
    /// <remarks>
    /// Half a minute is past the point where a card in the corner is being read and into the point where it is in the
    /// way, and a permanent delete waits it out, with a short grace behind it, before it reaches the mail server — so a
    /// bound above this would be a person asking their own mailbox to hold still for as long as they liked.
    /// </remarks>
    public const int LongestNotificationSeconds = 30;

    /// <summary>Gets whether a stated notification time is one this deployment accepts.</summary>
    /// <param name="seconds">The whole seconds a notification was asked to stand for.</param>
    /// <returns><see langword="true" /> when the value is within the bound, both ends included.</returns>
    /// <remarks>Asked at the boundary and answered with a refusal rather than clamped, because a client told its value was stored and given a different one is a screen that disagrees with the deployment about what somebody chose.</remarks>
    public static bool IsUsableNotificationTime(int seconds) =>
        seconds is >= ShortestNotificationSeconds and <= LongestNotificationSeconds;

    /// <summary>Gets what a person who has set nothing is answered with.</summary>
    /// <remarks>
    /// Telemetry on, because the switch withdraws a default this deployment already applies rather than granting one,
    /// and a stored answer that has never been written is not a refusal. The theme follows the machine, which is what
    /// the client resolves on the device before there is a session to read this at all. Tabs are off, because a person
    /// who has not asked for them is reading one message at a time. Marking read is on, because every mail client the
    /// user already uses does it and a client that leaves their read state behind is one they keep another beside.
    /// A conversation opens at the message it was opened at, because that is the message somebody came for and the
    /// history behind it is one control away. A message is read as the reduced text, because that is what this client
    /// has always drawn and the sender's own markup is a surface somebody asks for rather than one they are handed.
    /// The standing views are drawn, because they are a section of the tree somebody turns off rather than one they
    /// go looking for, and a deployment deriving nothing answers each of them as a list with nothing in it.
    /// A notification stands for five seconds, which is what the client's own toast has always stood for — so a
    /// deployment nobody has asked behaves as it did before this preference existed.
    /// </remarks>
    public static ClientPreferences Unset { get; } =
        new(true, ClientThemeChoice.System, false, true, false, ClientMessageView.Reduced, true, 5);
}

File: frontend/src/Client.Backend/src/clientPreferences.ts
// Copyright © 2026 Krzysztof Kasprowicz
// Licensed under the GNU Affero General Public License, Version 3. See LICENSE in the project root for license information.
// Project repository: https://github.com/Krzysztof318/MailFathom

import { failed, failureReasonForStatus, read, type ClientResult } from './failure';
import { asRecord } from './json';
import { headersFor, routeFor, type ClientSession } from './session';
import { spanned } from './telemetry';
import { send, type ClientResponse, type MailFathomTransport } from './transport';

/** The route the acting person's own client preferences are read at and written back to, relative to the client prefix. */
export const clientPreferencesRoute = '/preferences';

/** What a person may have the client painted in, as the deployment names it. */
export type ClientThemePreference = 'system' | 'light' | 'dark';

const themePreferences: readonly ClientThemePreference[] = ['system', 'light', 'dark'];

/**
 * Which of the three renderings a message opens on, as the deployment names it.
 *
 * `reduced` is the closed document tree the service reduces a body to and is what an unset preference reads as.
 * `cleaned` is that same tree with what the sender wrapped it in dropped, derived per open by a model that answers in
 * block indices. `embeddedHtml` is the sender's own markup, with everything that runs or reports removed.
 */
export type ClientMessageView = 'reduced' | 'cleaned' | 'embeddedHtml';

const messageViews: readonly ClientMessageView[] = ['reduced', 'cleaned', 'embeddedHtml'];

/**
 * What one person set about their own client, held on the deployment so it follows them between machines.
 *
 * Every preference is answered whether or not it was ever set, so a screen renders one shape rather than one per
 * combination of what happens to be stored. The document is closed and carries no version: a write states all of it,
 * and the last write wins.
 */
export interface ClientPreferences {
    readonly telemetryEnabled: boolean;
    readonly theme: ClientThemePreference;
    readonly openMailInTabs: boolean;

    /** Whether opening a message marks it read on the person's own mail server, across every account they read. */
    readonly markReadOnOpen: boolean;

    /** Whether opening a conversation draws every message in it rather than the one it was opened at. */
    readonly expandWholeThread: boolean;

    /** Which of the three renderings an open message is drawn on. */
    readonly messageView: ClientMessageView;

    /** Whether the folder tree carries the standing views of what a derivation read in the mail. */
    readonly aiFiltersShown: boolean;

    /**
     * How long one of the client's own notifications stands, in whole seconds.
     *
     * It is bounded by {@link shortestNotificationSeconds} and {@link longestNotificationSeconds}, and the deployment
     * refuses a value outside that rather than clamping it — so a client that sent one is told, instead of drawing a
     * setting the deployment quietly disagrees with.
     */
    readonly notificationSeconds: number;
}

/** The shortest a notification may be asked to stand for, which is the deployment's own bound. */
export const shortestNotificationSeconds = 1;

/** The longest a notification may be asked to stand for, which is the deployment's own bound. */
export const longestNotificationSeconds = 30;

/**
 * What somebody who has set nothing is answered with, which is also what stands in until an answer arrives.
 *
 * It restates the deployment's own unset answer rather than deriving one, so a client with nothing read yet draws the
 * same screen the first answer will confirm instead of one that changes under the reader for no reason they caused.
 */
export const unsetClientPreferences: ClientPreferences = {
    telemetryEnabled: true,
    theme: 'system',
    openMailInTabs: false,
    markReadOnOpen: true,
    expandWholeThread: false,
    messageView: 'reduced',
    aiFiltersShown: true,
    notificationSeconds: 5,
};

/**
 * The most of one preferences answer this package reads before refusing it.
 *
 * The document is eight scalars, so this is far above anything the deployment will legitimately send and far below
 * anything worth buffering. It is the same order the write route bounds its request body at, for the same reason:
 * what the bound guards against is an answer that was never a preferences document.
 */
export const longestPreferencesAnswer = 4_096;

/** Reads what the signed-in person set about their own client, answering an expected failure as a value. */
export function readClientPreferences(
    session: ClientSession,
    transport: MailFathomTransport,
): Promise<ClientResult<ClientPreferences>> {
    return spanned(`GET ${clientPreferencesRoute}`, async () => {
        return answerOf(
            await send(transport, {
                method: 'GET',
                path: routeFor(session, clientPreferencesRoute),
                headers: headersFor(session),
                longestAnswer: longestPreferencesAnswer,
            }),
        );
    });
}

/**
 * States the whole document, and answers with what is now stored.
 *
 * The whole of it rather than the part that changed, because that is what the route accepts: a preference the body
 * omits is committed as its own unset answer rather than left at whatever the row held. A caller therefore sends back
 * what it last read with one value replaced, which is also why nothing here merges anything.
 */
export function writeClientPreferences(
    session: ClientSession,
    transport: MailFathomTransport,
    stated: ClientPreferences,
): Promise<ClientResult<ClientPreferences>> {
    return spanned(`POST ${clientPreferencesRoute}`, async () => {
        return answerOf(
            await send(transport, {
                method: 'POST',
                path: routeFor(session, clientPreferencesRoute),
                headers: { ...headersFor(session), 'Content-Type': 'application/json' },
                body: JSON.stringify(stated),
                longestAnswer: longestPreferencesAnswer,
            }),
        );
    });
}

// Both routes answer the stored document, so both are read the same way. A deployment that holds no record for the
// caller answers the write with 404, which arrives here as `unavailable` like any other status this package did not
// expect to succeed: there is nothing a screen does differently about it that it does not already do about a
// deployment that would not take the change.
function answerOf(response: ClientResponse | null): ClientResult<ClientPreferences> {
    if (response === null) {
        return failed('unavailable', null);
    }

    if (response.status !== 200) {
        return failed(failureReasonForStatus(response.status), response.status);
    }

    const preferences = parsePreferences(response.body);

    return preferences === null ? failed('unreadable', response.status) : read(preferences);
}

function parsePreferences(body: string): ClientPreferences | null {
    let parsed: unknown;

    try {
        parsed = JSON.parse(body);
    } catch {
        return null;
    }

    const record = asRecord(parsed);
    if (record === null) {
        return null;
    }

    const telemetryEnabled = record['telemetryEnabled'];
    const theme = record['theme'];
    const openMailInTabs = record['openMailInTabs'];
    const markReadOnOpen = record['markReadOnOpen'];
    const expandWholeThread = record['expandWholeThread'];
    const messageView = record['messageView'];
    const aiFiltersShown = record['aiFiltersShown'];
    const notificationSeconds = record['notificationSeconds'];

    if (!isNotificationTime(notificationSeconds)) {
        return null;
    }

    if (
        typeof telemetryEnabled !== 'boolean' ||
        typeof openMailInTabs !== 'boolean' ||
        typeof markReadOnOpen !== 'boolean' ||
        typeof expandWholeThread !== 'boolean' ||
        typeof aiFiltersShown !== 'boolean'
    ) {
        return null;
    }

    if (!isThemePreference(theme) || !isMessageView(messageView)) {
        return null;
    }

    return {
        telemetryEnabled,
        theme,
        openMailInTabs,
        markReadOnOpen,
        expandWholeThread,
        messageView,
        aiFiltersShown,
        notificationSeconds,
    };
}

function isThemePreference(value: unknown): value is ClientThemePreference {
    return typeof value === 'string' && themePreferences.includes(value as ClientThemePreference);
}

function isMessageView(value: unknown): value is ClientMessageView {
    return typeof value === 'string' && messageViews.includes(value as ClientMessageView);
}

/**
 * Whether that is a notification time this deployment will take, which is a whole number of seconds inside the bound.
 *
 * Checked here as well as at the deployment's own boundary, for the reason every field on this surface is checked: an
 * answer is untrusted input, and a value outside the bound would reach a screen as a notification that never goes or
 * one nobody can read. A whole number, because the preference is stated in seconds and a fraction is not one.
 *
 * Exported because the screen that offers the setting asks the same question before stating one, and a second reading
 * of the bound is how a client comes to send what the deployment refuses.
 */
export function isNotificationTime(value: unknown): value is number {
    return (
        typeof value === 'number' &&
        Number.isInteger(value) &&
        value >= shortestNotificationSeconds &&
        value <= longestNotificationSeconds
    );
}
